### ISSUE TITLE
Lake frac_sno can remain 0 even when there is some snow addition from frost

### ISSUE THREAD

**Issue by billsacks** [ESCOMP/CTSM#827]
### Brief summary of bug

I have found that lake frac_sno can remain 0 even when there is a small amount of snow present, if this snow was added by frost rather than snowfall.

### General bug information

**CTSM version you are using:** ctsm1.0.dev074

**Does this bug cause significantly incorrect results in the model's science?** No: This does affect the model evolution, but I expect the impact to be very small.

**Configurations affected:** All

### Details of bug

The issue is that dew addition to the snow pack happens after frac_sno has been updated. For lakes, this is in LakeHydrologyMod. For non-lake columns, if there is not an explicit snow pack, then snow dew / sublimation is added to / subtracted from the top soil layer rather than h2osno_no_layers; but for lakes, this snow dew / sublimation is added to / subtracted from h2osno_no_layers, causing this issue. In theory, we should possibly also adjust frac_sno when there is dew / sublimation from an explicit-layer snow pack (currently I don't think that adjustment happens for either lake or non-lake columns), but that probably doesn't make a big difference.

Although this probably doesn't cause a big problem scientifically, it causes weirdness in the code because it means that we can have a non-zero snow pack but still 0 frac_sno. (Aside: this probably explains why I ran into trouble in ctsm1.0.dev074 and needed to avoid applying the lake frac_sno correction to newer initial conditions files; if we fix this issue, it's possible that we could back out some of the changes in ctsm1.0.dev074 to distinguish between new vs. old initial conditions, but it also doesn't hurt to keep those changes in place, and that might be safest.) With the way the frac_sno calculation is done, it will remain 0 until there is actually a snowfall event. I'm mainly concerned with the possibility that repeated dew-on-snow (without any real snowfall) could lead to the formation of an explicit-layered snow pack; if frac_sno were still 0 at that point, that could cause trouble once I have changed frac_sno_eff to be 0 if frac_sno is 0.

### Important details of your setup / configuration so we can reproduce the bug

I found this issue via answer changes in a few tests when I changed frac_sno_eff to be 0 if frac_sno is 0. Examples of tests illustrating this issue are: `ERP_D_Ld5.f19_g17.I2000Clm50BgcCruGs.cheyenne_intel.clm-default`, `ERP_Ld3.f09_g17.I1850Clm50BgcCropCru.cheyenne_intel.clm-ciso`, and `ERP_Ly3_P72x2.f10_f10_musgs.IHistClm50BgcCrop.cheyenne_intel.clm-cropMonthOutput`.

**Comment by billsacks** [comment]
My plan is to fix this by re-calling the routine UpdateSnowDepthAndFrac after calculating dew and sublimation on the snow pack. When I do this, I will also need an explicit new snow bulk density. I guess I'll keep the same logic as is currently there:

https://github.com/ESCOMP/CTSM/blob/837ad0dce955b09816ca8b8d5637f4545d308799/src/biogeophys/LakeHydrologyMod.F90#L330-L334

i.e., if there is already snow present, take the bulk density from the present snow; if not, assume a hard-coded bulk density (250 currently).

**Comment by billsacks** [comment]
This was my first attempt (diffs on top of ctsm1.0.dev074). **Note that I have NOT carefully reviewed this yet.**

```diff
diff --git a/src/biogeophys/LakeHydrologyMod.F90 b/src/biogeophys/LakeHydrologyMod.F90
index ff61eb5b..2d963e0e 100644
--- a/src/biogeophys/LakeHydrologyMod.F90
+++ b/src/biogeophys/LakeHydrologyMod.F90
@@ -23,6 +23,7 @@ module LakeHydrologyMod
   use clm_varpar           , only : nlevsno, nlevgrnd, nlevsoi
   use ColumnType           , only : col                
   use PatchType            , only : patch                
+  use filterColMod         , only : filter_col_type, col_filter_from_filter_and_logical_array
   use atm2lndType          , only : atm2lnd_type
   use AerosolMod           , only : aerosol_type
   use EnergyFluxType       , only : energyflux_type
@@ -116,14 +117,18 @@ subroutine LakeHydrology(bounds, &
     real(r8) :: fracsnow(bounds%begp:bounds%endp)               ! frac of precipitation that is snow
     real(r8) :: fracrain(bounds%begp:bounds%endp)               ! frac of precipitation that is rain
     real(r8) :: qflx_evap_soi_lim                               ! temporary evap_soi limited by top snow layer content [mm/s]
-    real(r8) :: h2osno_temp                                     ! temporary h2osno [kg/m^2]
+    real(r8) :: h2osno_before_update(bounds%begc:bounds%endc)   ! temporary h2osno before update [kg/m^2]
     real(r8) :: h2osno_total(bounds%begc:bounds%endc)           ! total snow water (mm H2O)
     real(r8) :: sumsnowice(bounds%begc:bounds%endc)             ! sum of snow ice if snow layers found above unfrozen lake [kg/m&2]
+    real(r8) :: snow_dew_minus_sub(bounds%begc:bounds%endc)     ! time step integrated dew minus sublimation onto snow [kg/m^2]
+    real(r8) :: zeros(bounds%begc:bounds%endc)                  ! 0 for columns that need it
     logical  :: unfrozen(bounds%begc:bounds%endc)               ! true if top lake layer is unfrozen with snow layers above
     real(r8) :: heatrem                                         ! used in case above [J/m^2]
     real(r8) :: heatsum(bounds%begc:bounds%endc)                ! used in case above [J/m^2]
     real(r8) :: snowmass                                        ! liquid+ice snow mass in a layer [kg/m2]
     real(r8) :: snowcap_scl_fct                                 ! temporary factor used to correct for snow capping
+    logical  :: needs_frac_sno_update(bounds%begc:bounds%endc)  ! whether a given column needs to have its frac_sno updated
+    type(filter_col_type) :: needs_frac_sno_update_filter       ! column filter: whether a given column needs to have its frac_sno updated
     real(r8), parameter :: snow_bd = 250._r8                    ! assumed snow bulk density (for lakes w/out resolved snow layers) [kg/m^3]
     ! Should only be used for frost below.
     !-----------------------------------------------------------------------
@@ -179,7 +184,8 @@ subroutine LakeHydrology(bounds, &
          h2osno_no_layers     => b_waterstate_inst%h2osno_no_layers_col        , & ! Output: [real(r8) (:)   ]  snow that is not resolved into layers (kg/m2)
          snowice              =>  b_waterdiagnostic_inst%snowice_col           , & ! Output: [real(r8) (:)   ]  average snow ice lens                   
          snowliq              =>  b_waterdiagnostic_inst%snowliq_col           , & ! Output: [real(r8) (:)   ]  average snow liquid water               
-         h2osoi_ice           =>  b_waterstate_inst%h2osoi_ice_col        , & ! Output: [real(r8) (:,:) ]  ice lens (kg/m2)                      
+         int_snow             =>  b_waterstate_inst%int_snow_col               , & ! Input:  [real(r8) (:)   ]  integrated snowfall (mm H2O)
+         h2osoi_ice           =>  b_waterstate_inst%h2osoi_ice_col        , & ! Output: [real(r8) (:,:) ]  ice lens (kg/m2)
          h2osoi_liq           =>  b_waterstate_inst%h2osoi_liq_col        , & ! Output: [real(r8) (:,:) ]  liquid water (kg/m2)                  
          h2osoi_vol           =>  b_waterstate_inst%h2osoi_vol_col        , & ! Output: [real(r8) (:,:) ]  volumetric soil water [m3/m3]         
          
@@ -269,6 +275,12 @@ subroutine LakeHydrology(bounds, &
 
     ! Calculate sublimation and dew, adapted from HydrologyLake and Biogeophysics2.
 
+    do fc = 1, num_lakec
+       c = filter_lakec(fc)
+       needs_frac_sno_update(c) = .false.
+    end do
+
+
     do fp = 1,num_lakep
        p = filter_lakep(fp)
        c = pcolumn(p)
@@ -324,19 +336,82 @@ subroutine LakeHydrology(bounds, &
           end if
 
           ! Update snow pack for dew & sub.
+          !
+          ! The following code currently assumes a single patch per column
 
-          h2osno_temp = h2osno_no_layers(c)
-          h2osno_no_layers(c) = h2osno_no_layers(c) + (-qflx_sub_snow(p)+qflx_dew_snow(p))*dtime
-          if (h2osno_temp > 0._r8) then
-             snow_depth(c) = snow_depth(c) * h2osno_no_layers(c) / h2osno_temp
-          else
-             snow_depth(c) = h2osno_no_layers(c)/snow_bd !Assume a constant snow bulk density = 250.
+          snow_dew_minus_sub(c) = (-qflx_sub_snow(p)+qflx_dew_snow(p))*dtime
+          h2osno_before_update(c) = h2osno_no_layers(c)
+          if (snow_dew_minus_sub(c) > 0._r8) then
+             ! NOTE(wjs, 2019-10-24) We currently don't adjust frac_sno for sublimation -
+             ! only for dew/frost.  If we change that logic, then we'd probably also need
+             ! to change the arguments to the UpdateSnowDepthAndFrac call (currently we
+             ! set newsnow = snow_dew_minus_sub; that won't work right if dew - sub < 0).
+             ! However, note that UpdateSnowDepthAndFrac seems to assume that the
+             ! passed-in h2osno_total has already been adjusted for any melt (but not for
+             ! accumulation); so we'd have to do something like: pass in h2osno_total
+             ! equal to the amount that has had sublimation subtracted but not dew added,
+             ! set snowmelt equal to sublimation, set newsnow equal to dew.
+             needs_frac_sno_update(c) = .true.
+             zeros(c) = 0._r8
+             if (h2osno_before_update(c) > 0._r8) then
+                ! Assume that frost has the same bulk density as the existing snow
+                bifall(c) = snow_depth(c) / h2osno_before_update(c)
+             else
+                ! Assume a constant bulk density of dew/frost on snow
+                bifall(c) = snow_bd
+             end if
+             ! TODO(wjs, 2019-10-24) Should int_snow also be added in this case?
           end if
-
+          h2osno_no_layers(c) = h2osno_no_layers(c) + snow_dew_minus_sub(c)
           h2osno_no_layers(c) = max(h2osno_no_layers(c), 0._r8)
+          if (snow_dew_minus_sub(c) < 0._r8) then
+             ! Decrease snow depth to maintain the same snow bulk density as before. Note
+             ! that any snow depth increase due to snow_dew_minus_sub > 0 will be handled
+             ! in the below call to UpdateSnowDepthAndFrac.
+             !
+             ! NOTE(wjs, 2019-10-24) If we changed the logic for needs_frac_sno_update so
+             ! that we also did a frac_sno update when snow_dew_minus_sub < 0, then we
+             ! might want to remove the snow_depth correction here. However, it's possible
+             ! that it would still be correct to update snow_depth here, since it isn't
+             ! updated with melt/sublimation in UpdateSnowDepthAndFrac (see
+             ! https://github.com/ESCOMP/CTSM/issues/828).
+             snow_depth(c) = snow_depth(c) * h2osno_no_layers(c) / h2osno_before_update(c)
+             if (h2osno_no_layers(c) <= 0._r8) then
+                frac_sno(c) = 0._r8
+                frac_sno_eff(c) = 0._r8
+             end if
+          end if
        end if
     end do
 
+    needs_frac_sno_update_filter = col_filter_from_filter_and_logical_array( &
+         bounds = bounds, &
+         num_orig = num_lakec, &
+         filter_orig = filter_lakec, &
+         logical_col = needs_frac_sno_update(bounds%begc:bounds%endc))
+
+    ! NOTE(wjs, 2019-10-24) Note that we currently only call this if snow_dew_minus_sub >
+    ! 0. If we change that, then we probably also need to change the arguments below,
+    ! since it won't work right for newsnow (currently equal to snow_dew_minus_sub) to be
+    ! negative. Also note that we pass in zeros for snowmelt to avoid triggering the melt
+    ! portion of this routine. (See also the other 2019-10-24 notes above.)
+    call scf_method%UpdateSnowDepthAndFrac( &
+         ! Inputs
+         bounds = bounds, &
+         num_c = needs_frac_sno_update_filter%num, &
+         filter_c = needs_frac_sno_update_filter%indices, &
+         lun_itype_col = col%lun_itype(begc:endc), &
+         urbpoi = col%urbpoi(begc:endc), &
+         h2osno_total = h2osno_before_update(begc:endc), &
+         snowmelt = zeros(begc:endc), &
+         int_snow = int_snow(begc:endc), &
+         newsnow = snow_dew_minus_sub(begc:endc), &
+         bifall = bifall(begc:endc), &
+         ! Outputs
+         snow_depth = snow_depth(begc:endc), &
+         frac_sno = frac_sno(begc:endc), &
+         frac_sno_eff = frac_sno_eff(begc:endc))
+
     ! patch averages must be done here -- BEFORE SNOW CALCULATIONS AS THEY USE IT.
     ! for output to history tape and other uses
     ! (note that pft2col is called before LakeHydrology, so we can't use that routine
```

However, I feel this introduces a lot of complexity for little scientific benefit. It also seems to carry assumptions about how UpdateSnowDepthAndFrac works, which could be fragile moving forward (e.g., if someone wanted to introduce another alternative for snow cover fraction method) - for example, assuming that it's safe to call that routine multiple times in the time step as long as snowmelt is set to 0 the second time. It also assumes that the parameterization used for new snow is applicable to frost. And there is the weird aspect that we update frac_sno when frost accumulates on a very thin (0-layer) snow pack, but we do NOT update frac_sno when frost accumulates on a thicker snow pack.

So now I'm thinking I will NOT use those changes, but instead will put in place a simpler fix for this problem. In particular, I plan to just set frac_sno to a very small but non-zero value when frost initiates a snow pack on a lake column.

**Comment by billsacks** [comment]
I have put in place a simpler (although admittedly somewhat arbitrary) fix in 30d8402c.
